crypto/x509/v3_purp.c

```diff
@@ -398,30 +417,34 @@ int x509v3_cache_extensions(X509 *x)
     }
     ERR_set_mark();
 
+    /* Cache the SHA1 digest of the cert */
     if (!X509_digest(x, EVP_sha1(), x->sha1_hash, NULL))
-            x->ex_flags |= EXFLAG_INVALID;
+        x->ex_flags |= EXFLAG_INVALID;
+    /*
+     * Note that the cert is marked invalid also on internal malloc failure
+     * or on failure of EVP_MD_fetch(), potentially called by X509_digest().
+     */
 
     /* V1 should mean no extensions ... */
     if (X509_get_version(x) == 0)
         x->ex_flags |= EXFLAG_V1;
 
     /* Handle basic constraints */
+    x->ex_pathlen = -1;
     if ((bs = X509_get_ext_d2i(x, NID_basic_constraints, &i, NULL)) != NULL) {
         if (bs->ca)
             x->ex_flags |= EXFLAG_CA;
         if (bs->pathlen != NULL) {
+            /*
+             * the error case !bs->ca is checked by check_chain_extensions()
+             * in case ctx->param->flags & X509_V_FLAG_X509_STRICT
```

Comment:
I'm not sure why you removed the check here and made it only part of the strict check, making OpenSSL tolerate more invalid certificates.

I'm not sure why OpenSSL allowed a non-CA certificate to have a pathlen of 0, since that doesn't make sense, the last CA in the chain has a pathlen of 0.

Reply:
Indeed the original form of the check was wrong - I guess whoever implemented this did not understand the semantics of pathlen == 0.

I moved the improved check to `check_chain_extensions()` for consistency with the other checks.
As far as I've learned doing this PR, `x509v3_cache_extensions()` basically does only those checks required for a cert to be usable in the crypto lib, while the non-essential checks are done in  `check_chain_extensions()`, where many of them are only done if strict checking is requested.

Reply:
You could argue that a function that does caching shouldn't even try to mark the certificate as invalid.

Reply:
I agree, yet on the other hand maybe the caller expects the SHA1 hash to be present when `x509v3_cache_extensions()` returns success, i.e., does not mark the cert is invalid. Yet this and many further such details were not documented, while this PR at least adds some basic documentation in `doc/internal/man3/x509v3_cache_extensions.pod`.